api/v1alpha1/linodevpc_webhook.go

```diff
@@ -0,0 +1,266 @@
+	defer cancel()
+
+	return nil, r.validateLinodeVPC(ctx, &defaultLinodeClient)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *LinodeVPC) ValidateUpdate(old runtime.Object) (admission.Warnings, error) {
+	linodevpclog.Info("validate update", "name", r.Name)
+
+	// TODO(user): fill in your validation logic upon object update.
+	return nil, nil
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *LinodeVPC) ValidateDelete() (admission.Warnings, error) {
+	linodevpclog.Info("validate delete", "name", r.Name)
+
+	// TODO(user): fill in your validation logic upon object deletion.
+	return nil, nil
+}
+
+func (r *LinodeVPC) validateLinodeVPC(ctx context.Context, client LinodeClient) error {
+	var errs field.ErrorList
+
+	if err := r.validateLinodeVPCSpec(ctx, client); err != nil {
+		errs = slices.Concat(errs, err)
+	}
+
+	if len(errs) == 0 {
+		return nil
+	}
+	return apierrors.NewInvalid(
+		schema.GroupKind{Group: "infrastructure.cluster.x-k8s.io", Kind: "LinodeVPC"},
+		r.Name, errs)
+}
+
+func (r *LinodeVPC) validateLinodeVPCSpec(ctx context.Context, client LinodeClient) field.ErrorList {
+	var errs field.ErrorList
+
+	if err := validateRegion(ctx, client, r.Spec.Region, field.NewPath("spec").Child("region"), LinodeVPCCapability); err != nil {
+		errs = append(errs, err)
+	}
+	if err := r.validateLinodeVPCSubnets(); err != nil {
+		errs = slices.Concat(errs, err)
+	}
+
+	if len(errs) == 0 {
+		return nil
+	}
+	return errs
+}
+
+func (r *LinodeVPC) validateLinodeVPCSubnets() field.ErrorList {
+	var (
+		errs    field.ErrorList
+		builder netipx.IPSetBuilder
+		cidrs   = &netipx.IPSet{}
+		labels  = []string{}
+	)
+
+	for i, subnet := range r.Spec.Subnets {
+		var (
+			label     = subnet.Label
+			labelPath = field.NewPath("spec").Child("Subnets").Index(i).Child("Label")
+			ip        = subnet.IPv4
+			ipPath    = field.NewPath("spec").Child("Subnets").Index(i).Child("IPv4")
+		)
+
+		// Validate Subnet Label
+		if err := validateVPCLabel(label, labelPath); err != nil {
+			errs = append(errs, err)
+		} else if slices.Contains(labels, label) {
+			errs = append(errs, field.Invalid(labelPath, label, "must be unique among the vpc's subnets"))
+		} else {
+			labels = append(labels, label)
+		}
+
+		// Validate Subnet IP Address Range
+		cidr, ferr := validateSubnetIPv4CIDR(ip, ipPath)
+		if ferr != nil {
+			errs = append(errs, ferr)
+			continue
+		}
+		if cidrs.Overlaps(cidr) {
+			errs = append(errs, field.Invalid(ipPath, ip, "range must not overlap with other subnets on the same vpc"))
+		}
+		var err error
+		builder.AddSet(cidr)
+		if cidrs, err = builder.IPSet(); err != nil {
+			return append(field.ErrorList{}, field.InternalError(ipPath, fmt.Errorf("build ip set: %w", err)))
+		}
+	}
+
+	if len(errs) == 0 {
+		return nil
+	}
+	return errs
+}
+
+// TODO: Replace the OpenAPI schema validation for .metadata.name.
```

Comment:
can you elaborate a little on what this TODO is for?

Reply:
The VPC Subnet label schema is essentially the same as the one for the VPC label itself. We previously [implemented the latter](https://www.github.com/linode/cluster-api-provider-linode/pull/219/commits/cc5e93415b446b2ca257681b0beac87e37622d73#diff-7613e13b49df95927633cbc1b5f25a84eafc7028e0cd6f5e45161e054d6a133e) via the Kubernetes Validation Rules feature onto the CustomResourceDefinitions definitions themselves. Syntactically, it might be better to remove those and consolidate all our custom (non-Kubernetes) validations in one place (the Go API resource definitions under `api/`).